src/common/json.cc

```diff
@@ -527,15 +525,118 @@ Json JsonReader::ParseObject() {
 }
 
 Json JsonReader::ParseNumber() {
-  std::string substr = raw_str_.substr(cursor_.Pos(), kMaxNumLength);
-  size_t pos = 0;
+  // Adopted from sajson with some simplifications and small optimizations.
```

Comment:
Have you considered using `dmlc::strtof()` to parse floating-point numbers?
https://github.com/dmlc/dmlc-core/blob/808f485387f9a03f78fa9f1159f387d0d91b7a28/include/dmlc/strtonum.h#L99

Some differences:
* `dmlc::strtof()` contains more checks for overflow and underflow conditions
* `dmlc::strtof()` supports NAN and INF
* `dmlc::strtof()` avoids the use of `exp10` and `exp` by repeating `*10`: 
https://github.com/dmlc/dmlc-core/blob/808f485387f9a03f78fa9f1159f387d0d91b7a28/include/dmlc/strtonum.h#L184-L199

I think your implementation is reasonable. I'll defer to you to choose between your implementation or `dmlc::strtof()`.

Reply:
@hcho3 The problem is we also support integer parsing.  It's not in json specification but added by most of implementations.  Also we have lots of integers in tree node index.  So I need a function that can determine whether the string is an integer or a float.

Reply:
@trivialfis Got it. In that case, let's just keep `JsonReader::ParseNumber()` as it is.

In the future, when we decide to move your JSON serializer into dmlc-core, we can discuss how to combine the logic in `JsonReader::ParseNumber()` and `dmlc::ParseFloat()` for code de-duplication.